core/src/main/scala/kafka/admin/ZkSecurityMigrator.scala

```diff
@@ -0,0 +1,214 @@
+import org.apache.zookeeper.CreateMode
+import org.apache.zookeeper.data.Stat
+import org.apache.zookeeper.KeeperException
+import org.apache.zookeeper.KeeperException.Code
+import org.apache.zookeeper.ZooDefs.Ids
+import org.apache.zookeeper.ZooKeeper
+
+import scala.collection.JavaConverters._
+import scala.collection._
+import scala.concurrent._
+
+object ZkSecurityMigrator extends Logging {
+  val counter = new AtomicInteger(0)
+  var exception: ZkException = null
+  val workQueue = new LinkedBlockingQueue[Runnable]
+  val threadExecutionContext = 
+    ExecutionContext.fromExecutor(new ThreadPoolExecutor(1,
+                                                         Runtime.getRuntime().availableProcessors(),
+                                                         5000,
+                                                         TimeUnit.MILLISECONDS,
+                                                         workQueue))
+
+  def setAclsRecursively(zkUtils: ZkUtils, path: String) {
+    counter.incrementAndGet()
+    zkUtils.zkConnection.getZookeeper.getChildren(path, false, GetChildrenCallback, zkUtils)
+  }
+
+  object GetChildrenCallback extends ChildrenCallback {
+      def processResult(rc: Int,
+                        path: String,
+                        ctx: Object,
+                        children: java.util.List[String]) {
+        val list = if(children == null)
+          null
+        else
+          children.asScala.toList
+        val zkUtils: ZkUtils = ctx.asInstanceOf[ZkUtils]
+        val zkHandle = zkUtils.zkConnection.getZookeeper
+        
+        Code.get(rc) match {
+          case Code.OK =>
+            // Set ACL for each child
+            if(list != null) 
+              Future {
+                val childPathBuilder = new StringBuilder
+                for(child <- list) {
+                  childPathBuilder.clear
+                  childPathBuilder.append(path)
+                  childPathBuilder.append("/")
+                  childPathBuilder.append(child)
+
+                  val childPath = childPathBuilder.toString
+                  counter.incrementAndGet()
+                  zkHandle.setACL(childPath, ZkUtils.DefaultAcls(true), -1, SetACLCallback, ctx)
+                  zkHandle.getChildren(childPath, false, GetChildrenCallback, ctx)
+                }
+                counter.decrementAndGet()
+              }(threadExecutionContext)
+          case Code.CONNECTIONLOSS =>
+            setAclsRecursively(zkUtils, path)
+          case Code.NONODE =>
+            warn("Node is gone, it could be have been legitimately deleted: %s".format(path))
+          case Code.SESSIONEXPIRED =>
+            // Starting a new session isn't really a problem, but it'd complicate
+            // the logic of the tool, so we quit and let the user re-run it.
+            error("ZooKeeper session expired while changing ACLs")
+            exception = throw ZkException.create(KeeperException.create(rc))
```

Comment:
I don't understand what's happening here. Why are we assigning even though we are throwing an exception?